docs/server datastorage.md

```diff
@@ -0,0 +1,173 @@
+    ]
+}
+```
+
+### Queue (FIFO) \ Stack (FILO) 
+Queues can be implemented using lists and this will still be safe aslong as all producers only append to the end of the list and consumers only remove the first or last element in the list. To append to a list we would use the `add` operation with an array value, this will append all elements in the provided array to the list. In order to `peek` any elemment in the list, you would unfortunately have to retrieve the full list by its key. To remove the first or last element of a list we can use the `pop` opperation.
+
+Adding elements to the list is easy, just a `Set`-`add` operation with an array as value:
+```json
+{
+    "cmd": "Set",
+    "key": "SomeQueueName",
+    "default": [],
+    "operations": [
+        {"operation": "add", "value": ["whatever value"] },
+    ]
+}
+```
+To pop an element in the queue we would use a `Set`-`pop` operation, with a value of `0` for queues and `1` for stacks
+```json
+{
+    "cmd": "Set",
+    "key": "SomeQueueName",
+    "default": [],
+    "on_error": "ignore",
+    "tag": "4CDF4226-C7D2-4242-8636-D16FA9EA2016",
+    "want_reply": true,
+    "operations": [
+        {"operation": "pop", "value": 0 },
+    ]
+}
+```
+Than inside the `SetReply` check by tag if this was done as a result to you pop-ing it, if thats the case you find the pop-ed element in the `original_value`. 
+
+### Locking
+This is not an actual used example but if you want an even more complex thread safe way you can implement using the techniques aboves. it is possible to use a `pop` operation on specific dictionary keys along with a tagging your `Set` command to see if you client can obtain a lock, for example:
+```json
+{
+    "cmd": "Set",
+    "key": "MyStoreAvailableKeys",
+    "default": { "Key1": true, "Key2": true },
+    "on_error": "ignore",
+    "tag": "2c14b824-f274-40e2-ab88-e3ce643683c1",
+    "want_reply": true,
+    "operations": [
+        {"operation": "pop", "value": "Key1" },
+    ]
+}
+```
+Than inside the `SetReply` check by looking in the `original_value` and 'value' if `Key1` got removed and by tag if this was done as a result to you pop-ing it, if thats all true, you know you now got the lock and can do whatever you want to the key `MyStoreAvailableKeys:Key1`. and when your done with it, you write back `Key1` to the `MyStoreAvailableKeys` using an update.
+```json
+{
+    "cmd": "Set",
+    "key": "MyStoreAvailableKeys",
+    "default": {},
+    "operations": [
+        {"operation": "update", "value": { "Key1": true }}
+    ]
+}
+```
+Ofcourse this is not perfect, its only safe if all clients trying to alter `MyStoreAvailableKeys:Key1` respect this locking mechanism, and if the client who obtained the lock disconnects or otherwise fails to release it you need to somehow timeout that lock, like you could have a separated dict storing the last timestamps for each key a lock was obtained.

```

Comment:
```suggestion
Of course this is not perfect; it's only safe if all clients trying to alter `Key1` respect this locking mechanism, and if the client who obtained the lock disconnects or otherwise fails to release it you may want additional mechanisms such as timeouts on the lock to prevent it from being permanently lost.

```